src/Authentication/Authentication/Cmdlets/Environment/SetMgEnvironment.cs

```diff
@@ -0,0 +1,106 @@
+﻿// ------------------------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
+// ------------------------------------------------------------------------------
+
+namespace Microsoft.Graph.PowerShell.Authentication.Cmdlets
+{
+    using Microsoft.Graph.PowerShell.Authentication.Extensions;
+    using System;
+    using System.Globalization;
+    using System.Linq;
+    using System.Management.Automation;
+
+    /// <summary>
+    /// Sets a Microsoft Graph environment to the settings file..
+    /// </summary>
+    [Cmdlet(VerbsCommon.Set, "MgEnvironment", SupportsShouldProcess = true)]
+    [OutputType(typeof(GraphEnvironment))]
+    public class SetMgEnvironment : PSCmdlet
+    {
+        [Parameter(Position = 0, Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = "The environment name.")]
+        [ValidateNotNullOrEmpty]
+        public string Name { get; set; }
+
+        [Parameter(Position = 1, Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "The base URL for Azure AD endpoint to get access tokens for Microsoft Graph endpoint.")]
+        [ValidateNotNullOrEmpty]
+        [Alias("AzureADUrl")]
+        public string AzureADEndpoint { get; set; }
+
+        [Parameter(Position = 2, Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "The service root endpoint for Microsoft Graph.")]
+        [ValidateNotNullOrEmpty]
+        [Alias("GraphUrl")]
+        public string GraphEndpoint { get; set; }
+
+        protected override void BeginProcessing()
+        {
+            base.BeginProcessing();
+            ValidateParameters();
+        }
+
+        protected override void ProcessRecord()
+        {
+            base.ProcessRecord();
+
+            if (ShouldProcess("updating environment", Name))
+            {
+                if (GraphEnvironment.BuiltInEnvironments.Keys.Any((k) => string.Equals(k, Name, StringComparison.CurrentCultureIgnoreCase)))
+                {
+                    throw new InvalidOperationException(string.Format(CultureInfo.CurrentCulture,
+                            "Cannot change built-in environment {0}.", Name));
+                }
+
+                GraphSettings settings = this.GetContextSettings();
+                if (!settings.TryGetEnvironment(Name, out IGraphEnvironment newEnvironment))
+                {
+                    // Create default environment if it doesn't exist.
+                    newEnvironment = new GraphEnvironment {
```

Comment:
this seems like it may just end up causing issues if a user incorrectly put in an environment name and then just one parameter, you get a mix of a global and another setting that's not likely the right combination.  may be better just to throw here.